File: apps/local_demo/static/hardware.js
"use strict";

// All hardware mutations are explicit POSTs. Polls only read server state.
const hardware = {
  ticket: null, state: null, ready: false, pending: false, polling: false,

  async request(path, payload) {
    const response = await fetch(path, payload === undefined ? {} : {
      method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify(payload),
    });
    const result = await response.json();
    if (!response.ok) {
      if (result.hardware) this.render(result.hardware);
      throw new Error(result.error || "Hardware request failed.");
    }
    return result;
  },

  clearPreview() {
    this.ticket = null;
    this.ready = false;
    $("#send-readiness").textContent = "Calculate a fresh preview before sending.";
    this.buttons();
  },

  setPreview(ticket) {
    this.ticket = ticket;
    this.ready = false;
    this.check();
  },

  buttons() {
    const state = this.state;
    const busy = this.pending || state?.operation_pending || state?.connection === "connecting";
    $("#send-device").disabled = !this.ticket || !this.ready || busy || state?.connection !== "connected";
    $("#connect-device").disabled = busy || !state?.available || state?.connection === "connected";
    $("#disconnect-device").disabled = busy || state?.connection !== "connected";
    $("#serial-port").disabled = busy || state?.connection === "connected";
    $("#refresh-ports").disabled = busy || !state?.available;
  },

  render(state) {
    this.state = state;
    if (state.connection === "connected" && state.port) $("#serial-port").value = state.port;
    $("#connection-state").textContent = !state.available ? "Hardware unavailable" :
      ({disconnected: "Disconnected", connecting: "Connecting", connected: "Connected"}[state.connection]);
    $("#hardware-message").textContent = state.message;
    $("#simulation-note").hidden = !state.simulated;
    $("#device-info").hidden = !state.capabilities;
    if (state.capabilities) {
      const identity = $("#device-identity");
      identity.replaceChildren();
      field(identity, "Device ID", state.device_id);
      field(identity, "Port", state.port);
      field(identity, "Duration", `${state.capabilities.min_duration}–${state.capabilities.max_duration} s`);
      field(identity, "Channels", state.capabilities.channels.map((item) => `${item.channel}: ${item.min_intensity}–${item.max_intensity}`).join("; "));
      $("#device-capabilities").textContent = JSON.stringify(state.capabilities, null, 2);
    }
    const execution = state.execution;
    $("#execution-info").hidden = !execution;
    if (execution) {
      $("#execution-duration").textContent = `Submitted plan duration: ${number.format(execution.duration)} s.`;
      $("#execution-heading").textContent = {sending: "Sending…", accepted: "Command accepted",
        rejected: "Command rejected", uncertain: "Execution uncertain"}[execution.state];
      const remaining = state.remaining_seconds;
      $("#execution-estimate").textContent = remaining > 0 ?
        `Estimated wait: ${Math.ceil(remaining)} s. Local estimate only; no device completion signal.` :
        execution.state === "accepted" ? "Estimated duration elapsed. Completion has not been confirmed by the device." :
        execution.state === "uncertain" ? "The command may have run. Check the LEDs before reconnecting or sending again." :
        execution.state === "rejected" ? "The device returned an error. No automatic retry." : "Waiting for a device response…";
      $("#execution-response").textContent = execution.response ? JSON.stringify(execution.response, null, 2) :
        execution.detail || "No response received yet.";
    }
    this.buttons();
  },

  async check() {
    const ticket = this.ticket;
    if (!ticket || this.pending) return;
    this.ready = false;
    this.buttons();
    if (this.state && !this.state.available) {
      $("#send-readiness").textContent = "Hardware unavailable. Preview remains available.";
      return;
    }
    try {
      const result = await this.request("/api/hardware/check", ticket);
      if (this.ticket !== ticket || this.pending) return;
      this.ready = result.ready;
      $("#send-readiness").textContent = result.message;
    } catch (error) {
      if (this.ticket !== ticket || this.pending) return;
      $("#send-readiness").textContent = error.message;
    }
    this.buttons();
  },

  async refresh() {
    if (this.polling || this.pending) return;
    this.polling = true;
    try {
      const state = await this.request("/api/hardware/status");
      if (!this.pending) { this.render(state); await this.check(); }
    } catch (_) {
      this.ready = false;
      $("#hardware-message").textContent = "Cannot reach the local server. Connection and execution state are unknown. No automatic reconnect or resend.";
      $("#connection-state").textContent = "Connection unknown";
      this.buttons();
    } finally {
      this.polling = false;
    }
  },

  async action(action) {
    if (this.pending) return;
    if (action === "send" && (!this.ready || !this.ticket)) return;
    const ticket = this.ticket;
    this.pending = true;
    this.ready = false;
    $("#hardware-error").hidden = true;
    if (action === "connect") $("#connection-state").textContent = "Connecting";
    if (action === "send") {
      this.clearPreview();
      document.querySelectorAll("#load-demo, #import-file, #duration, #duration-form button, #policy-picker input").forEach((item) => { item.disabled = true; });
      $("#send-readiness").textContent = "Sending the displayed plan…";
    }
    this.buttons();
    try {
      const payload = action === "connect" ? {port: $("#serial-port").value} : action === "send" ? ticket : {};
      this.render(await this.request(`/api/hardware/${action}`, payload));
    } catch (error) {
      $("#hardware-error").textContent = error instanceof TypeError ?
        "Request outcome unknown. Check the device and server status. Do not automatically repeat the action." : error.message;
      $("#hardware-error").hidden = false;
    } finally {
      this.pending = false;
      document.querySelectorAll("#load-demo, #import-file, #duration, #duration-form button, #policy-picker input").forEach((item) => { item.disabled = false; });
      if (action === "send") $("#send-readiness").textContent = "Recalculate for a new explicit send.";
      this.buttons();
      await this.refresh();
    }
  },
};

$("#refresh-ports").addEventListener("click", async () => {
  $("#refresh-ports").disabled = true;
  $("#port-message").textContent = "Listing serial ports…";
  try {
    const result = await hardware.request("/api/hardware/ports");
    const list = $("#serial-ports");
    list.replaceChildren();
    for (const item of result.ports) {
      const option = node("option", `${item.port} — ${item.description}`);
      option.value = item.port;
      list.append(option);
    }
    $("#port-message").textContent = result.ports.length ? result.message : "No ports found. Connect the USB device, refresh, or enter a port manually.";
  } catch (error) {
    $("#port-message").textContent = error.message;
  } finally { hardware.buttons(); }
});
$("#connect-device").addEventListener("click", () => hardware.action("connect"));
$("#disconnect-device").addEventListener("click", () => hardware.action("disconnect"));
$("#send-device").addEventListener("click", () => hardware.action("send"));
hardware.refresh();
setInterval(() => hardware.refresh(), 1000);

File: apps/local_demo/hardware.py
"""Explicit hardware actions, isolated from graph loading and preview mapping."""

from contextlib import contextmanager
from copy import deepcopy
from dataclasses import asdict
import importlib.util
import threading
import time

EXPECTED_DEVICE_ID = "opensmell-esp32-led-3ch-001"
INSTALL_HELP = 'PySerial is not installed. From the repository root, run: python -m pip install -e ".[serial]". Preview remains available.'


class HardwareError(ValueError):
    def __init__(self, message, status=409):
        super().__init__(message)
        self.status = status


def serial_available():
    return importlib.util.find_spec("serial") is not None


def list_ports():
    # Enumeration only: never open ports to identify devices.
    from serial.tools import list_ports as ports
    return [{"port": item.device, "description": item.description} for item in ports.comports()]


def open_transport(port):
    import serial
    from opensmell.experimental.serial_device_transport import SerialDeviceTransport

    # Own the serial handle before SDK startup/reset preparation, so even a
    # preparation failure closes it. Same settings as the physical CLI demo.
    connection = serial.Serial(port=port, baudrate=115200, timeout=2.0, write_timeout=2.0)
    try:
        return SerialDeviceTransport(
            port=port, baudrate=115200, timeout=2.0, startup_delay=2.0,
            discard_initial_input=True, serial_instance=connection,
        )
    except BaseException:
        connection.close()
        raise


class RecordedTransport:
    """Keep the last reply for a faithful error diagnostic, without retrying."""
    def __init__(self, transport):
        self.transport = transport
        self.reply = None

    def exchange(self, message):
        self.reply = None
        self.reply = self.transport.exchange(message)
        return self.reply


def device_error_reply(reply):
    # Reuse SDK parsing and error construction to distinguish a valid device
    # rejection from malformed/absent replies. The adapter still interprets OK.
    from opensmell.experimental.device_protocol import error_response, loads_message
    try:
        parsed = loads_message(reply)
        expected = error_response(parsed.get("code"), parsed.get("message"))
        if all(parsed.get(key) == value for key, value in expected.items()):
            return parsed
    except (TypeError, ValueError):
        pass
    return None


class HardwareService:
    def __init__(self, *, factory=open_transport, available=serial_available,
                 ports=list_ports, clock=time.monotonic, simulated=False):
        self._factory, self._available, self._ports = factory, available, ports
        self._clock = clock
        self._simulated = simulated
        self._operation = threading.Lock()
        self._state_lock = threading.RLock()
        self._state = "disconnected"
        self._port = None
        self._adapter = self._transport = self._recorded = None
        self._message = "Select a port and click Connect."
        self._execution = None
        self._busy_until = 0.0
        self._closed = False

    @contextmanager
    def operation(self):
        if not self._operation.acquire(blocking=False):
            raise HardwareError("A hardware operation is already in progress. No second command was sent.")
        try:
            if self._closed:
                raise HardwareError("The server is shutting down.")
            yield
        finally:
            self._operation.release()

    def status(self):
        with self._state_lock:
            available = self._available()
            return {
                "available": available, "simulated": self._simulated,
                "connection": self._state, "port": self._port,
                "device_id": self._adapter.device_id if self._adapter else None,
                "capabilities": asdict(self._adapter.capabilities) if self._adapter else None,
                "message": self._message if available else INSTALL_HELP,
                "execution": deepcopy(self._execution),
                "remaining_seconds": max(0.0, self._busy_until - self._clock()),
                "operation_pending": self._operation.locked(),
            }

    def ports(self):
        if not self._available():
            return {"available": False, "ports": [], "message": INSTALL_HELP}
        try:
            return {"available": True, "ports": self._ports(), "message": "Choose a listed port or enter its name."}
        except Exception as exc:
            raise HardwareError(f"Cannot list serial ports. You can enter a port manually. Details: {exc}", 503) from exc

    def _require_idle_execution(self):
        remaining = self._busy_until - self._clock()
        if remaining > 0:
            raise HardwareError(
                f"Wait for the expected execution window ({remaining:.1f} s remaining). "
                "This is a local estimate, not a device completion signal."
            )

    def _close_transport(self):
        transport = self._transport
        self._adapter = self._recorded = self._transport = None
        self._state = "disconnected"
        if transport:
            try:
                transport.close()
            except Exception as exc:
                return f" Serial close failed: {exc}. Check the USB connection."
        return ""

    def connect(self, port):
        if not isinstance(port, str) or not port.strip() or len(port) > 256 or any(c in port for c in "\r\n\0"):
            raise HardwareError("Enter a valid serial port name, such as COM3.", 400)
        with self.operation():
            with self._state_lock:
                if not self._available():
                    raise HardwareError(INSTALL_HELP, 503)
                if self._adapter:
                    raise HardwareError("Already connected. Disconnect before choosing another port.")
                self._require_idle_execution()
                self._port = port.strip()
                self._state = "connecting"
                self._message = "Connecting; waiting for device startup, identity and capabilities…"
            try:
                from opensmell.experimental.protocol_device_adapter import ProtocolDeviceAdapter
                transport = self._factory(self._port)
                with self._state_lock:
                    self._transport = transport
                recorded = RecordedTransport(transport)
                adapter = ProtocolDeviceAdapter(recorded)
                if adapter.device_id != EXPECTED_DEVICE_ID:
                    raise HardwareError(f"Unexpected device: {adapter.device_id!r}. Expected {EXPECTED_DEVICE_ID}.")
                with self._state_lock:
                    self._adapter, self._recorded = adapter, recorded
                    self._state = "connected"
                    self._message = "Connected. Identity and capabilities received from the device."
            except Exception as exc:
                with self._state_lock:
                    cleanup = self._close_transport()
                    self._message = (f"Connection failed: {exc}. Check the selected port, USB cable and firmware; "
                                     "close any serial monitor. No render command was sent." + cleanup)
                raise HardwareError(self._message, 503) from exc
        return self.status()

    def _require_plan(self, plan):
        if not self._adapter:
            raise HardwareError("Connect the device before sending.")
        self._require_idle_execution()
        if not plan.commands:
            raise HardwareError("An empty plan can be previewed but cannot be sent to hardware.")
        try:
            self._adapter.capabilities.require_plan(plan)
        except ValueError as exc:
            raise HardwareError(f"Plan incompatible with the device: {exc}. Nothing was sent.") from exc

    def check(self, previews, ticket):
        # Lock order matches send: preview state then hardware state. No I/O.
        with previews.current(ticket) as entry, self._state_lock:
            if self._operation.locked():
                raise HardwareError("A hardware operation is in progress.")
            self._require_plan(entry["plan"])
        return {"ready": True, "message": "Ready to send"}

    def send(self, previews, ticket):
        with self.operation():
            with previews.current(ticket) as entry, self._state_lock:
                self._require_plan(entry["plan"])
                plan = deepcopy(entry["plan"])
                # Atomic freshness check/consumption. Even a failed exchange
                # cannot reuse this ticket and duplicate a possible command.
                entry["used"] = True
                self._execution = {"state": "sending", "duration": plan.duration,
                                   "preview_id": ticket["preview_id"], "response": None}
                self._message = "Sending the previewed plan…"
            try:
                response = self._adapter.render(plan)
            except Exception as exc:
                rejection = device_error_reply(self._recorded.reply)
                with self._state_lock:
                    if rejection:
                        self._execution.update(state="rejected", response=rejection)
                        self._message = f"Device rejected the command: {rejection['code']}: {rejection['message']}. No automatic retry."
                    else:
                        self._execution.update(state="uncertain", detail=str(exc))
                        self._busy_until = self._clock() + plan.duration
                        cleanup = self._close_transport()
                        self._message = (f"Execution uncertain: the command may have reached the device. {exc}. "
                                         "The connection was closed. Check the LEDs before reconnecting; no automatic retry." + cleanup)
                raise HardwareError(self._message, 502) from exc
            with self._state_lock:
                self._execution.update(state="accepted", response=response)
                self._busy_until = self._clock() + plan.duration
                self._message = "Command accepted. The device does not report completion or LED state."
        return self.status()

    def disconnect(self):
        with self.operation(), self._state_lock:
            cleanup = self._close_transport()
            self._message = "Disconnected. Closing the link does not stop an accepted command." + cleanup
        return self.status()

    def close(self):
        # Server shutdown waits for any bounded serial exchange before closing.
        with self._operation, self._state_lock:
            self._closed = True
            self._close_transport()
